crates/sui-graphql-rpc/src/metrics.rs

```diff
@@ -1,15 +1,209 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use prometheus::{register_histogram_with_registry, Histogram, Registry};
-
-#[derive(Clone, Debug)]
-pub struct RequestMetrics {
-    pub(crate) input_nodes: Histogram,
-    pub(crate) output_nodes: Histogram,
-    pub(crate) query_depth: Histogram,
-    pub(crate) query_payload_size: Histogram,
-    pub(crate) _db_query_cost: Histogram,
+use std::sync::Arc;
+
+use async_graphql::{PathSegment, ServerError};
+use prometheus::{
+    register_histogram_vec_with_registry, register_histogram_with_registry,
+    register_int_counter_vec_with_registry, Histogram, HistogramVec, IntCounterVec, Registry,
+};
+use std::fmt::Write;
+
+use crate::error::code;
+
+#[derive(Clone)]
+pub(crate) struct Metrics {
+    pub db_metrics: Arc<DBMetrics>,
+    pub request_metrics: Arc<RequestMetrics>,
+}
+
+impl Metrics {
+    pub(crate) fn new(registry: &Registry) -> Self {
+        let db_metrics = DBMetrics::new(&registry);
+        let request_metrics = RequestMetrics::new(&registry);
+
+        Self {
+            db_metrics: Arc::new(db_metrics),
+            request_metrics: Arc::new(request_metrics),
+        }
+    }
+
+    /// Updates the DB related metrics (latency, error, success)
+    pub(crate) fn observe_db_data(&self, db_latency: f64, succeeded: bool) {
+        self.db_metrics
+            .db_fetch_latency
+            .with_label_values(&["latency", "db"])
+            .observe(db_latency);
+        if succeeded {
+            self.db_metrics
+                .db_num_fetches
+                .with_label_values(&["success"])
+                .inc();
+        } else {
+            self.db_metrics
+                .db_num_fetches
+                .with_label_values(&["error"])
+                .inc();
+        }
+    }
+
+    /// Use this function to increment the number of errors due to bad input
+    pub(crate) fn inc_num_bad_input_errors(&self, label: &str) {
+        self.request_metrics
+            .num_bad_input_errors_by_path
+            // TODO when measuring cost related errors, there's no path
+            // not 100% how to label such an error?
+            .with_label_values(&[label])
+            .inc();
+    }
+
+    /// Use this function to increment the number of errors per path, and labels
+    /// them as either internal server error, bad user input, or any other error.
+    /// The path is the same as the path that the user sees.
+    pub(crate) fn inc_errors(&self, errors: Vec<ServerError>) {
+        for err in errors {
+            if let Some(ext) = err.extensions {
+                if let Some(code) = ext.get("code") {
+                    if let async_graphql_value::Value::String(val) = code.clone().into_value() {
+                        if val == code::INTERNAL_SERVER_ERROR {
```

Comment:
What are your thoughts on splitting out these nested ifs into

let Some(ext) = err.extensions else continue

let Some(code) = ...

Yes we have a lot of continue but it's easier on the eyes since we don't have such deep nesting. You can also chain Some, like let Some(Some... 

I'm guessing code is a GqlValue and you're converting it into a ConstValue

But then a bit confused about the val == code::INTERNAL_SERVER_ERROR, how does that work?

Reply:
`ext.get("code")` returns a `&ConstValue`, so need to call `into_value()` on it to be able to match on `Value::String`. Otherwise it complains that it cannot do a comparison between `Value` and `String`.

Reply:
i dont follow why we are comparing val against itself:

if let async_graphql_value::Value::String(val) =
                                code.clone().into_value()
                            {
                                if val == code::INTERNAL_SERVER_ERROR {

unless I'm not understanding it correctly? val is from code.clone().into_value(). then we compare val against code::INTERNAL_SERVER_ERROR? 

oh. is code from error::code? If so, can we avoid using the same word?

Reply:
Sorry for confusing you. The high level goal is to match on the error code that we're setting in the `error.rs` file. I will change `code` to `error_code`, and import the error codes constants so that we have more like this or better maybe I can do some maps to get the set error code as a string:
```
            if let Some(ext) = err.extensions {
                if let Some(error_code) = ext.get("code") {
                    if let async_graphql_value::Value::String(val) = error_code.clone().into_value() {
                        if val == INTERNAL_SERVER_ERROR {
```

Reply:
> ext.get("code") returns a &ConstValue, so need to call into_value() on it to be able to match on Value::String. Otherwise it complains that it cannot do a comparison between Value and String.

Oddly enough, when I try the following code snippet, I do not get the error. Also, I'd think that one would be able to match a reference to a ConstValue to async_graphql::Value::String just like a reference to any other enum. 

    if let Some(code) = ext.get("code") {
        if let async_graphql::Value::String(val) = code {
            if val == code::BAD_USER_INPUT {
                println!("oops");
            }
        }
    }

Reply:
Admittedly, I did not give it a lot of thought. I'll look at it again and try to simplify.